Add swagger API file (#7)
iotaledger/iota-core: documentation/APIs/core-models.md

# Core Models

This document defines core models for IOTA V3 protocol.

* [Block](#block)
  * [Block ID](#block-id)
  * [Block Syntactic Validation Rules](#block-syntactic-validation-rules)
* [Slot Index](#slot-index)
* [Slot Commitment](#slot-commitment)
  * [Slot Commitment ID](#slot-commitment-id)
* [Payloads](#payloads)
  * [Tagged Data Payload](#tagged-data-payload)
  * [Transaction](#transaction)

>   * [Outputs and Unlock blocks](#outputs-and-unlock-blocks)

Reviewer: ```suggestion
  * [Outputs and Unlocks](#outputs-and-unlocks)
```
Should `blocks` be removed? I think we did it in Stardust everywhere except for the actual blocks

            <th>Name</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
          <tr>
            <td>Block ID</td>
            <td>ByteArray[40]</td>
            <td>The Block ID of the parent.</td>
          </tr>
        </table>
      </details>
    </td>
  </tr>
  <tr>
  <tr>
    <td>Shallow Like Parents Count</td>
    <td>uint8</td>
    <td>The number of blocks that are directly referenced to adjust opinion.</td>
  </tr>
  <tr>
    <td valign="top">Shallow Like Parents <code>anyOf</code></td>
    <td colspan="2">
      <details>
        <summary>Parent</summary>
        <table>
          <tr>
            <th>Name</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
          <tr>
            <td>Block ID</td>
            <td>ByteArray[40]</td>
            <td>The Block ID of the parent.</td>
          </tr>
        </table>
      </details>
    </td>
  </tr>
  <tr>
    <td>Issuer ID</td>
    <td>ByteArray[32]</td>
    <td>The issuer identifier</td>
  </tr>
  <tr>
    <td>Issuing Time</td>
    <td>uint64</td>
    <td>The time the block is issued. It's a Unix-like timestamp in nanosecond.</td>
  </tr>
  <tr>

>     <td valign="top">Slot Commitment <code>oneOf</code></td>

Reviewer 2: Unnecessary `oneOf`?

        <blockquote>
          References another directly approved block.
        </blockquote>
        <table>
          <tr>
            <th>Name</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
          <tr>
            <td>Block ID</td>
            <td>ByteArray[40]</td>
            <td>The Block ID of the parent.</td>
          </tr>
        </table>
      </details>
    </td>
  </tr>
  <tr>
  <tr>
    <td>Shallow Like Parents Count</td>
    <td>uint8</td>
    <td>The number of blocks that are directly referenced to adjust opinion.</td>
  </tr>
  <tr>
    <td valign="top">Shallow Like Parents <code>anyOf</code></td>
    <td colspan="2">
      <details>
        <summary>Parent</summary>
        <table>
          <tr>
            <th>Name</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
          <tr>
            <td>Block ID</td>
            <td>ByteArray[40]</td>
            <td>The Block ID of the parent.</td>
          </tr>
        </table>
      </details>
    </td>
  </tr>
  <tr>
    <td>Issuer ID</td>
    <td>ByteArray[32]</td>
    <td>The issuer identifier</td>
  </tr>
  <tr>
    <td>Issuing Time</td>

>     <td>int64</td>

Reviewer 2: Does this need to be a signed integer?

    <td>The time the block is issued.</td>
  </tr>
  <tr>

>     <td>Slot Commitment Length</td>

Reviewer 2: What is this length for? A `Slot Commitment` size doesn't look dynamic.

Reviewer 2: @jkrvivian I see that this was closed without answering or addressing it, can you let me know what's the reason for this field please?

Reviewer 3: Yeah, I removed this field, it's a fixed-size and not an optional structure, so the structure is serialized anyway.
I also double-check with the bytes from serix, we don't need the length.

    <td>uint32</td>
    <td>The length of serialized Slot Commitment structure.</td>
  </tr>
  <tr>
    <td valign="top">Slot Commitment <code>oneOf</code></td>
    <td colspan="2">
      <details open="true">
        <summary>Slot Commitment</summary>
        <blockquote>
          Slot Commitment is an object that contains a summary of a slot. More descriptions are in Slot Commitment section.
        </blockquote>
      </details>
    </td>
  </tr>
  <tr>
    <td>Latest Finalized Slot</td>

>     <td>int64</td>

Reviewer 2: Does this need to be a signed integer?

    <td>The slot index of latest finalized slot.</td>
  </tr>
  <tr>
    <td>Payload Length</td>
    <td>uint32</td>
    <td>The length of the following payload in bytes. A length of 0 means no payload will be attached.</td>
  </tr>
  <tr>

>     <td valign="top">Payload <code>oneOf</code></td>

Reviewer 2: Should be `optOneOf` ?

          </tr>
          <tr>
            <td>Payload Bytes</td>
            <td>ByteArray</td>
            <td>A sequence of serialized bytes, where the structure depends on `Payload Type`.</td>
          </tr>
        </table>
      </details>
    </td>
  </tr>
  <tr>
    <td valign="top">Signature <code>oneOf</code></td>
    <td colspan="2">
      <details open="true">
        <summary>Generic Signature</summary>
        <table>
          <tr>
            <th>Name</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
          <tr>
            <td>Signature Type</td>
            <td>uint8</td>
            <td>
              The type of the signature. It will instruct the node how to parse the fields that follow.
            </td>
          </tr>
          <tr>
            <td>Signature Bytes</td>
            <td>ByteArray</td>
            <td>A sequence of serialized bytes, where the structure depends on `Signature Type`.</td>
          </tr>
        </table>
      </details>
    </td>
  </tr>
  <tr>
    <td>Nonce</td>
    <td>uint64</td>
    <td>The nonce which lets this block fulfill the PoW requirement.</td>
  </tr>
</table>

### Block ID

> Block ID denotes an identifier of a block, with type `ArrayBytes[40]`. It is calculated as the following steps, using BLAKE2b-256 hash function:

Reviewer 2: Out of curiosity, has blake3 been considered for protocol v3?

Reviewer 3: I don't think so, we don't have this discussion around at the moment.

Reviewer 2: Ok, thanks

* `content` is the serialized block **without** signature and nonce.

> * `slot_index` is the slot index of the `Issuing Time` of the block.

Reviewer 2: Is that the index in the Slot Commitment struct?

Reviewer 3: No, it's a different one. I'll clarify that in the context!

> * `signatureBytes` is the serialized Ed25519 signature.

Reviewer 2: Signature is supposed to be generic so maybe don't mention Ed25519 here?

Calculation:
1. `content_hash` = hash(`content`)

> 2. `id` = hash(Concat(`content_hash`, signatureBytes))

Reviewer 3: We're having a discussion on this but might have an answer by next Tuesday. :cry:

3. And finally, `BlockID` = Concat(`id`, `slot_index`) 

The string format of Block ID is hexadecimal encoding of Block ID with `0x` prefix.

### Block Syntactic Validation Rules

* Block size must not exceeds `MaxBlockSize` bytes, currently is `32768`.
* `Protocol Version` must match the `Protocol Version` config parameter of the node.

> * Must have at least **1** `Strong Parents`.

Reviewer 2: Just to be sure, this means we can have no weak and no shallow like parents? To disambiguate, I'd give an actual allowed range for each of them.

* `Strong Parents`, `Weak Parents`, `Shallow Like Parents` must not exceeds `BlockMaxParents` (currently is **8**) per type.

> * `Strong Parents`, `Weak Parents`, `Shallow Like Parents` of parents must be lexically ordered.

Reviewer 2: Remove `of parents` ?

* `Strong Parents`, `Weak Parents`, `Shallow Like Parents` must not have duplicates in each list.

> * `Weak Parents` must be disjunct to the rest of the parents.

Reviewer 2: So no weak parent at all should also be in strong or shallow like?

* `Payload Type` must match one of the values described under Payloads.
  Data Fields must be correctly parsable in the context of the `Payload Type`.
  The payload itself must pass syntactic validation.

> * `Signature` must be valid, the input of a signature is concatation of Issuing Time bytes, Commitment ID bytes and Block bytes (without Signature and Nonce) by order.

Reviewer 2: Where is `Commitment ID` from?

Reviewer 2: is that `Slot Commitment ID` ?

> * `Signature` must be valid, the input of a signature is concatation of Issuing Time bytes, Commitment ID bytes and Block bytes (without Signature and Nonce) by order.

Reviewer 2: Isn't `Issuing Time` included twice in the signature then ?

Reviewer 2: Same as Commitment ID?

Reviewer 3: Firstly, Block bytes are not the input of the Signature, it's `content_hash` instead, already fixed it in the doc. :sweat_smile:
 
And yes, we'll need the Issuing Time explicitly to perform validation in slot commitment chain fork.

Reviewer 2: What do you mean it's been fixed? I still see this:
> Signature must be valid, the input of a signature is concatation of Issuing Time bytes, Commitment ID bytes and Block bytes (without Signature and Nonce) by order.

Reviewer 2: Where is `Slot Commitment ID` from? I assume that's *not* the one from within the Block itself?

Reviewer 3: It's the ID of Slot Commitment in a Block. I'll add this description.

Reviewer 2: Ok so it's `Previous Slot Commitment ID`

Reviewer 3: Nope, it's the ID calculated from the Slot Commitment structure of the block.

Quote from the Slot Commitment ID section:

Slot Commitment ID is calculated as the following steps, using BLAKE2b-256 hash function:

* `content` is the serialized slot commitment.
* `slot_index` is the slot index of the slot commitment (`Index` from the Slot Commitment Structure).

Caculation:
1. `content_hash` = hash(`content`)
2. And finally, `CommitmentID` = Concat(`content_hash`, `slot_index`)

Reviewer 3: Let me add a Block Signature section to go through the details.

* There must be no trailing bytes after all block fields have been parsed.

## Slot Index

Timeline is divided into slots, and each slot has a corresponding slot index, which is a `int64`.
To calculate the slot index of a timestamp, `genesisTimestamp` and the duration of a slot are needed.
The slot index of timestamp `ts` is `(ts - genesisTimestamp)/duration + 1`.

> ## Slot Commitment

Reviewer 3: Slot Commitment is part of the `/block/issuance` API, clients can get it from the node, with no need to compute themselves.

Commitment is an object that contains a summary of a slot, and it is linked to the commitment of the previous slot, which forms a commitment chain. 

Multiple sparse merkle trees are managed by per slot, which are:
* *Tangle tree*: All accepted blocks within a slot.
* *State mutation tree*: All accepted transactions within a slot.
* *Activity tree*: All active accounts within a slot.
* *State tree*: All accepted created/consumed outputs within a slot.

> * *Mana tree*: Mana of all accounts at the end of slot.

Reviewer 2: ```suggestion
* *Mana tree*: Mana of all accounts at the end of a slot.
```

When a slot is committable, new commitment is generated and will be included in the new issued blocks.

**Note** that more trees will be added in the future, but this will only change the calculation of `Root ID`, the Slot Commitment structure stays the same.

The following table describes the serialization of a Slot Commitment:

<table>
  <tr>
    <th>Name</th>
    <th>Type</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>Index</td>

>     <td>int64</td>

Reviewer 2: signed integer?

    <td>
      The slot index of this commitment. It is calculated based on genesis timestamp and the duration of a slot.
    </td>
  </tr>
  <tr>
    <td>Previous Slot Commitment ID</td>
    <td>ByteArray[40]</td>
    <td>The commitment ID of the previous slot.</td>
  </tr>
  <tr>
    <td>Root ID</td>
    <td>ByteArray[32]</td>
    <td>

>       A BLAKE2b-256 hash of concating multiple sparse merkle tree roots of a slot.

Reviewer 2: ```suggestion
      A BLAKE2b-256 hash of concatenating multiple sparse merkle tree roots of a slot.
```

      <i>Root ID = hash( Concat( hash(Concat(Tangle, TX)), hash(Concat(State, Mana)) ) )</i>  
    </td>
  </tr>
  <tr>
    <td>Cumulative Weight</td>

>     <td>int64</td>

Reviewer 2: signed integer?

    <td>The sum of previous slot commitment cumulative weight and weight of issuers of accepted blocks within this slot. It is just an indication of "committed into" this slot, and can not strictly be used for evaluating the switching of a chain.</td>
  </tr>
</table>

### Slot Commitment ID

Slot Commitment ID denotes an identifier of a slot commitment, with type `ArrayBytes[40]`. It is calculated as the following steps, using BLAKE2b-256 hash function:

* `content` is the serialized slot commitment.
* `slot_index` is the slot index of the slot commitment.

Caculation:
1. `content_hash` = hash(`content`)

> 2. And finally, `CommitmentID` = Concat(`content_hash`, `slot_index`) 

Reviewer 2: For my understanding, what's the point of concatenating the slot_index here?

Author: It's for data locality, we do similar thing for blockID, this way you always know where to look for the commitment/block in the database given it is organized per slot based on its ID

Reviewer 3: It's a new type of identifier, from which we can derive the slot index directly from an ID.

The string format of Slot Commitment ID is hexadecimal encoding of Slot Commitment ID with `0x` prefix.

## Payloads
The following table lists all currently specified payloads that can be part of a block. 

<table>
    <tr>
        <th>Name</th>
        <th>Payload Type</th>
        <th>Description</th>
    </tr>
    <tr>
        <td>Tagged Data</td>
        <td>5</td>
        <td>A payload which holds a tag and associated data.</td>
    </tr>
    <tr>
        <td>Transaction</td>

>         <td>6</td>

Reviewer 2: Do we not need a new payload type since this transaction is not compatible with the stardust one?

Reviewer 3: New version is introduced in transaction essence type.

Reviewer 2: Ah yeah we keep the transaction but bump the essence, I see. Thanks!

          With Payload Type 5, more details are described in Tagged Data Payload section.
        </blockquote>
      </details>
      <details>
        <summary>Transaction Payload</summary>
        <blockquote>
          With Payload Type 6, more details are described in Transaction section.
        </blockquote>
      </details>
  </tr>
  <tr>
    <td valign="top">Signature <code>oneOf</code></td>
    <td colspan="2">
      <details>
        <summary>Ed25519 Signature</summary>
        <blockquote>
          With Signature Type 0, more details are described in Signature section.
        </blockquote>
      </details>
    </td>
    </td>
  </tr>
  <tr>
    <td>Nonce</td>
    <td>uint64</td>
    <td>The nonce which lets this block fulfill the PoW requirement.</td>
  </tr>
</table>

### Block Signature
This is how the signature is generated:

* `content` is the serialized block **without** signature and nonce.
* `slot_commitment_ID` is the identifier calculated from the Slot Commitment within the block, following steps described in [Slot Commitment ID](#slot-commitment-id) section.
* `Issuing Time` is the block issuing time.

Calculation:
1. `content_hash` = hash(`content`)

> 2. `data` = Concat(`Issuing Time`, `slot_commitment_ID`, `content_hash`)

Reviewer 3: We want to be able to prove the time without the block itself, so the signature verification works exactly the same as the block. That way you can prove the block existed, the slot commitment is the correct one and the issuing time is valid.
And this is currently used to perform validation in slot commitment chain fork.

@thibault-martinez here's a more detailed response :stuck_out_tongue: 

I'll add it to the doc.

3. And sign with the Issuer PubKey, `Signature` = Sign(`data`) 

### Block ID
Block ID denotes an identifier of a block, with type `ArrayBytes[40]`. It is calculated as the following steps, using BLAKE2b-256 hash function:

* `content` is the serialized block **without** signature and nonce.
* `slot_index` is the slot index of the `Issuing Time` of the block.

>    **Note**: It's **not** the same slot index as in the Slot Commitment.

Reviewer 2: So it means that we can't deduce the BlockID just by looking at the Block?

Reviewer 3: We're having a discussion on this but might have an answer by next Wednesday. :cry:

    <td>uint64</td>
    <td>
      The slot index of this commitment. It is calculated based on genesis timestamp and the duration of a slot.
    </td>
  </tr>
  <tr>
    <td>Previous Slot Commitment ID</td>
    <td>ByteArray[40]</td>
    <td>The commitment ID of the previous slot.</td>
  </tr>
  <tr>
    <td>Roots ID</td>
    <td>ByteArray[32]</td>
    <td>
      A BLAKE2b-256 hash of concatenating multiple sparse merkle tree roots of a slot.
      <i>Root ID = hash( Concat( hash(Concat(Tangle, TX)), hash(Concat(State, Mana)) ) )</i>  
    </td>
  </tr>
  <tr>
    <td>Cumulative Weight</td>
    <td>uint64</td>
    <td>The sum of previous slot commitment cumulative weight and weight of issuers of accepted blocks within this slot. It is just an indication of "committed into" this slot, and can not strictly be used for evaluating the switching of a chain.</td>
  </tr>
</table>

### Slot Commitment ID

Slot Commitment ID denotes an identifier of a slot commitment, with type `ArrayBytes[40]`. It is calculated as the following steps, using BLAKE2b-256 hash function:

* `content` is the serialized slot commitment.
* `slot_index` is the slot index of the slot commitment.

> Caculation:

Reviewer 4: ```suggestion
Calculation:
```

        <summary>Transaction Payload</summary>
        <blockquote>
          With Payload Type 6, more details are described in Transaction section.
        </blockquote>
      </details>
  </tr>
  <tr>
    <td valign="top">Signature <code>oneOf</code></td>
    <td colspan="2">
      <details>
        <summary>Ed25519 Signature</summary>
        <blockquote>
          With Signature Type 0, more details are described in Signature section.
        </blockquote>
      </details>
    </td>
    </td>
  </tr>
  <tr>
    <td>Nonce</td>
    <td>uint64</td>
    <td>The nonce which lets this block fulfill the PoW requirement.</td>
  </tr>
</table>

### Block Signature
This is how the signature is generated:

* `content` is the serialized block **without** signature and nonce.
* `slot_commitment_ID` is the identifier calculated from the Slot Commitment within the block, following steps described in [Slot Commitment ID](#slot-commitment-id) section.
* `Issuing Time` is the block issuing time.

Calculation:
1. `content_hash` = hash(`content`)
2. `data` = Concat(`Issuing Time`, `slot_commitment_ID`, `content_hash`)
3. And sign with the Issuer PubKey, `Signature` = Sign(`data`) 

> Including `Issuing Time` explicitly in the signature makes the signature verification works exactly the same as a block. So one can prove the block existed, the content and Slot Commitment are correct, and the Issuing Time is valid.

Reviewer 4: ```suggestion
Including `Issuing Time` explicitly in the signature makes the signature verification work exactly the same as a block's. So one can prove that the block existed, the content and Slot Commitment are correct, and the Issuing Time is valid.
```
